modal/snapshot.py

```diff
@@ -19,27 +22,35 @@ class _SandboxSnapshot(_Object, type_prefix="sn"):
     the original Sandbox at the time the snapshot was taken.
     """
 
-    @staticmethod
-    async def from_id(sandbox_snapshot_id: str, client: Optional[_Client] = None):
+    @deprecate_aio_usage((2025, 12, 5), "SandboxSnapshot.from_id")
+    @classmethod
+    def from_id(
+        cls, sandbox_snapshot_id: str, client: Optional["modal.client.Client"] = None
+    ) -> typing_extensions.Self:
         """
         Construct a `SandboxSnapshot` object from a sandbox snapshot ID.
         """
-        # TODO: remove this - from_id constructor should not do io:
-        client = client or await _Client.from_env()
+        _client = cast(_Client, synchronizer._translate_in(client))
 
         async def _load(
             self: _SandboxSnapshot, resolver: Resolver, load_context: LoadContext, existing_object_id: Optional[str]
         ):
-            await load_context.client.stub.SandboxSnapshotGet(
+            # hydration doesn't actually do much apart from validating the existance of the id
+            # which is implicitly done by trying to start a sandbox from the snapshot as well
+            resp: api_pb2.SandboxSnapshotGetResponse = await load_context.client.stub.SandboxSnapshotGet(
                 api_pb2.SandboxSnapshotGetRequest(snapshot_id=sandbox_snapshot_id)
             )
+            self._hydrate(resp.snapshot_id, load_context.client, None)
 
         rep = "SandboxSnapshot()"
-        obj = _SandboxSnapshot._from_loader(_load, rep, load_context_overrides=LoadContext(client=client))
-        # TODO: should this be a _Object._new_hydrated instead?
-        obj._hydrate(sandbox_snapshot_id, client, None)
-
-        return obj
+        obj = _SandboxSnapshot._from_loader(
+            _load, rep, load_context_overrides=LoadContext(client=_client), hydrate_lazily=True
+        )
+        # Setting the object id directly is a bit hacky, but
+        # it avoids hydrating the object fully if it's going
+        # to be used only for its object id anyway
+        obj._object_id = sandbox_snapshot_id
```

Comment:
This was mostly so we don't add an extra and unnecessary `SandboxSnapshotGet` rpc by forcing actual hydration of this object in Sandbox.from_snapshot...

Reply:
I'm open to that, it makes sense as a requirement. I think it's less important to be incredibly strict about the cardinality of hydration states and more important that it's very clear when looking at the code (ideally both from the perspective of a human developer and the perspective of the type checker) what state the object is required to be in.